ssl/statem/statem_clnt.c

```diff
@@ -821,7 +821,10 @@ int tls_construct_client_hello(SSL *s)
         goto err;
     }
 
-    if ((sess == NULL) || (sess->ssl_version != s->version) ||
+    if ((sess == NULL) ||
+        /* Only destroy session which version is bigger than maximum supported.
+         */
+        ssl_version_cmp(s, sess->ssl_version, s->version) > 0 ||
```

Comment:
I think it is doomed to fail isn't it (as long as the server resumes the session)?? We are talking about a version locked method requesting a session for a different version; the server resuming that session and therefore using that different version, which the client will then abort the connection because it is not prepared to talk that version.